cloudinit/net/__init__.py

```diff
@@ -1139,6 +1139,13 @@ def has_url_connectivity(url_data: Dict[str, Any]) -> bool:
     return True
 
 
+def maybe_get_address(check_cb: Callable, address: str, **kwargs):
```

Comment:
I don't think this is quite what we want as we now add another level of indirection to all `is_ip*` functions to now invoke yet another function under the hood. It looks like the reason you are centralizing this function is to provide VMware datasource with one less call to `if net.is_ip_address(val): <then check link_local etc>`.  I think I'd prefer if we can extend `net.is_ip_address` to do this optional checking itself given a new parameter (no_link_local=True, or allow_unroutable=False). See more historic discussion here https://github.com/canonical/cloud-init/pull/953/files#r682164843

Reply:
> The logic was basically: "this looks a lot like a version of is_ip_addr() that returns the address if true, rather than just a boolean". The extra layer of indirection allows us to use the existing library logic in the vmware datasource.

Yeah I think this logic in my mind went the other way: "This logic is almost like is_ip_addr() and only needs modification for VMWare datasource, why impact all other network_validator wrapper functions for something that seems specific to just is_ip_addr plus additional validation constraints".


Honestly, I don't know if other datasources would use specific additional constraints in is_ip_address if we built it into that function.

>  [Option 3]: network_validator() [could] return the address (probably also rename the function to something better describing the new functionality)
I hadn't thought of that option either. 

My comment on this shouldn't be blocking as it feels a bit more like personal preference of mine more than significant performance optimization.

I think either stay the course with what you have currently or go with option 3, which could be just drop the `network_validator` function, migrate the applicable parts of docstr of `network_validator` into  `maybe_get_address`  and perform the bool in the `is_ip_*`.
That at least would cut down on the number of functions we have to look through to see what is happening.